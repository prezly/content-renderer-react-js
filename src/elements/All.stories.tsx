import React from 'react';
import type { Meta, Story } from '@storybook/react';
import { importAll } from '../dev/importAll';
import { Renderer } from '../Renderer';

export default {
    title: 'AllElements',
} as Meta;

export const AllElements: Story = () => {
    const storiesModules = importAll(require.context('./', true, /\.slate.stories.tsx$/));

    const children: JSX.Element[] = [];

    storiesModules.forEach((module) => {
        children.push(
            <Renderer
                nodes={{
                    type: 'heading-one',
                    children: [{ text: module.default.title }],
                }}
                defaultComponents
            />,
        );

        Object.keys(module).forEach((key) => {
            const Story = module[key];

            if (typeof Story === 'function') {
                children.push(
                    <Renderer
                        nodes={{
                            type: 'heading-two',
                            children: [{ text: key.replace(/([A-Z])/g, ' $1') }],
                        }}
                        defaultComponents
                    />,
                );
                children.push(<Story />);
            }
        });
    });

    return <>{children}</>;
};
