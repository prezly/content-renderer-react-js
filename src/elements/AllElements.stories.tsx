import React from 'react';
import type { Meta, Story } from '@storybook/react';
import createAsyncCallback from '@loki/create-async-callback';
import { importAll } from '../dev/importAll';
import { Renderer } from '../Renderer';

export default {
    title: 'AllElements',
} as Meta;

export const AllElements: Story = () => {
    const cb = createAsyncCallback();

    setTimeout(() => {
        cb();
    }, 1000);

    const storiesModules = importAll(require.context('./', true, /\.slate.stories.tsx$/));

    const children: JSX.Element[] = [];

    storiesModules.forEach((module) => {
        children.push(
            <Renderer
                nodes={[
                    {
                        type: 'heading-one',
                        children: [{ text: module.default.title }],
                    },
                ]}
            />,
        );

        Object.keys(module).forEach((key) => {
            const Story = module[key];

            if (typeof Story === 'function') {
                children.push(
                    <Renderer
                        nodes={{
                            type: 'heading-two',
                            children: [{ text: key.replace(/([A-Z])/g, ' $1') }],
                        }}
                    />,
                );

                children.push(<Story />);
            }
        });

        children.push(
            <Renderer
                nodes={[
                    {
                        type: 'divider',
                        children: [
                            {
                                text: '',
                            },
                        ],
                    },
                ]}
            />,
        );
    });

    return <>{children}</>;
};
