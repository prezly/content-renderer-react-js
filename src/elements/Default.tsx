import React from 'react';
import type { Node } from 'slate';
import { describeNode } from '../lib';

interface Props {
    node: Node;
}

export function Default({ node }: Props) {
    return (
        <div style={{ color: 'red' }}>
            <strong>[@prezly/content-renderer-react-js]</strong> Unknown node type encountered:
            <pre>{describeNode(node)}</pre>
        </div>
    );
}
