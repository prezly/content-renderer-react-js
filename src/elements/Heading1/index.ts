export { default } from './Heading1';
