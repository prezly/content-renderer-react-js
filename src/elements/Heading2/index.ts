export { default } from './Heading2';
