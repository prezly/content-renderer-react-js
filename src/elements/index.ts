export { default as Attachment } from './Attachment';
export { default as BulletedList } from './BulletedList';
export { default as Contact } from './Contact';
export { default as Divider } from './Divider';
export { default as Document } from './Document';
export { default as Embed } from './Embed';
export { default as Gallery } from './Gallery';
export { default as Heading1 } from './Heading1';
export { default as Heading2 } from './Heading2';
export { default as Image } from './Image';
export { default as Link } from './Link';
export { default as ListItem } from './ListItem';
export { default as ListItemText } from './ListItemText';
export { default as NumberedList } from './NumberedList';
export { default as Paragraph } from './Paragraph';
export { default as Quote } from './Quote';
