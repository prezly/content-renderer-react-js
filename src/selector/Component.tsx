import type { Node } from 'slate';
import type { ComponentType } from 'react';
import { invariant } from './lib';

interface ComponentProps<T> {
    node: T;
}

type Props<T extends Node, P extends ComponentProps<T>> = {
    match: (node: Node) => node is T;
    component: ComponentType<P & ComponentProps<T>>;
} & Omit<P, 'node'>;

export function Component<T extends Node, P extends ComponentProps<T>>(_: Props<T, P>) {
    invariant(
        false,
        `A <Component> is only ever to be used as the child of <Selector> element, ` +
            `never rendered directly. Please wrap your <Component> in a <Selector>.`,
    );
    return null;
}
