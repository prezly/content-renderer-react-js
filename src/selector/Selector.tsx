import { ComposedElement, type Node } from '@prezly/story-content-format';
import { Fragment, type ReactNode } from 'react';

import { createComponentsRenderersFromChildren } from './lib';

interface Props {
    nodes: Node | Node[];
    children: ReactNode;
}

export function Selector({ nodes, children }: Props) {
    const renderers = createComponentsRenderersFromChildren(children);

    function renderNodes(nodes: Node[]): ReactNode {
        return nodes.map((node, idx) => <Fragment key={idx}>{renderNode(node)}</Fragment>);
    }

    function renderNode(node: Node) {
        for (const { match, component: ComponentRenderer, ...extraProps } of renderers) {
            if (match(node)) {
                return (
                    <ComponentRenderer
                        {...extraProps}
                        node={node}
                        children={
                            ComposedElement.isComposedElement(node)
                                ? renderNodes(node.children as Node[])
                                : undefined
                        }
                    />
                );
            }
        }
        return null;
    }

    return <>{Array.isArray(nodes) ? renderNodes(nodes) : renderNode(nodes)}</>;
}
