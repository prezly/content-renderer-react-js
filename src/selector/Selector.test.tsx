import { DividerNode, DocumentNode, HeadingNode } from '@prezly/story-content-format';
import renderer from 'react-test-renderer';

import { Divider, Document, Heading } from '../elements';

import { Component } from './Component';
import { Selector } from './Selector';

const documentNode: DocumentNode = {
    type: DocumentNode.TYPE,
    version: '0.50',
    children: [
        {
            type: HeadingNode.Type.HEADING_ONE,
            children: [{ text: 'Hello world!' }],
        },
        {
            type: DividerNode.TYPE,
        },
    ],
};

describe('Selector', () => {
    it('should render <h1> for a heading and <hr> for a divider', () => {
        const tree = renderer
            .create(
                <Selector nodes={documentNode}>
                    <Component match={DividerNode.isDividerNode} component={Divider} />
                    <Component match={HeadingNode.isHeadingNode} component={Heading} />
                    <Component match={DocumentNode.isDocumentNode} component={Document} />
                </Selector>,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
