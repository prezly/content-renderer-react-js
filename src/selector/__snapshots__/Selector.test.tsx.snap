// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selector should render <h1> for a heading and <hr> for a divider 1`] = `
<section
  className="prezly-slate-document"
  data-version="0.50"
>
  <h1
    className="prezly-slate-heading prezly-slate-heading--heading-1 prezly-slate-heading--align-inherit"
  />
  <hr
    className="prezly-slate-divider"
  />
</section>
`;
